### DISCUSSION TITLE
Toggling element visibility causes noticeable slowdown and lags

### DISCUSSION THREAD

**Discussion by MatthewAry** [bcakmakoglu/vue-flow#1444] [General]
So I'm using Vue Flow to render components as nodes for a data graph. They are fairly complex in comparison to the examples shown on the docs. I've made it so I can hide all descendant nodes by clicking a button on a parent node. The action of hiding descendants is fast but unhiding them is very slow and locks up the main thread.
![image](https://github.com/bcakmakoglu/vue-flow/assets/157217/24d37ef2-f04f-4ee3-b9ac-5e8332c75d81)
When unhiding elements from the root node, the operation took over 100ms to complete. The flame graph on the perf monitor is pretty deep here.

Unhiding should be fairly quick. Would it be possible to do some kind of memoization of previously hidden elements instead of completely un-rendering them? That, or hide them using CSS? 

The hide operation is handled by Vue Flow itself so I don't really have control over it too much.

Also, I'm not able to use Vue Flow's built in node parent referencing capabilities because it affects the way layouts are computed in an undesirable way.

**Comment by bcakmakoglu** [comment]
The issue seems to be performance which could be anything really, since you haven't presented any repro or code-snippets I don't know what your application looks like and if you're not doing somehow causing some excessive re-renders unknowingly.
This sort of report *really needs a reproduction*.

Looking at the screenshot, re-mounting 5 nodes should not have any noticeable impact on performance whatsoever, even if they have a bit more complexity, which they don't look like they do in your img but who knows - could it be you're doing some heavy computations in those nodes? Is that necessary?

Anyway, I'm moving this to discussions since there's no actual bug described here; possibly a feature/refactor request for how visibility of nodes could be handled instead :)

**Comment by bcakmakoglu** [comment]
> Also, I'm not able to use Vue Flow's built in node parent referencing capabilities because it affects the way layouts are computed in an undesirable way.

[Check out the `Nested Nodes` example.](https://vueflow.dev/examples/nodes/nesting.html)

This feature is meant for these sort of graphs, where you use nested nodes.
VueFlow doesn't compute any "layout", what you're most likely referring to is the change of the relation of `node.position` from viewport to parent node.
When you use the `parentNode` option this relation changes as I mentioned, so a regular node with `{ x: 0, y: 0 }` would be top-left of the viewport (the big white background), given the viewport is at it's default position, whereas a node using `parentId` and `{ x: 0, y: 0 }` would be at the top-left corner of the specified parent node.

**Comment by MatthewAry** [comment]
Reproduction https://github.com/MatthewAry/vue-flow-1444

**Reply by bcakmakoglu** [reply]
Thx :) I'll give it a look

**Reply by bcakmakoglu** [reply]
From what I can tell what causes a significant drop in performance is this bit of code:
```ts
// useLayout.ts

    /**
     * This is a hacky solution to make the layout system work and be stable.
     */
    let stable = false
    let stableCount = 0
    const requiredStableIterations = 3
    let iterations = 0
    const maxIterations = 300

    while (stableCount < requiredStableIterations && iterations < maxIterations) {
      dagre.layout(dagreGraph)

      const newNodes = nodes.value.map((node) => {
        const nodeWithPosition = dagreGraph.node(node.id)
        const { data, ...rest } = node
        return {
          data,
          ...rest,
          targetPosition: isHorizontal ? Position.Left : Position.Top,
          sourcePosition: isHorizontal ? Position.Right : Position.Bottom,
          position: { x: nodeWithPosition.x, y: nodeWithPosition.y },
        }
      })

      stable = JSON.stringify(newNodes) === JSON.stringify(nodes.value)

      if (stable) {
        stableCount++
      }
      else {
        stableCount = 0
      }

      nodes.value = newNodes

      await nextTick()

      iterations++
    }
```

Timing the layout function
- with the code above: 1370 ms
- without the code above: 3 ms

What exactly is that piece of code about? Seems like it's very inefficient :(

Sidenote:
You're providing the vue flow instance using `provide(Flow, flow)` which is unnecessary.
`useVueFlow` already does this for you, if you call it inside of a custom node or something it will use the same store instance

**Reply by MatthewAry** [reply]
Layout shouldn't be getting invoked when unhiding nodes. The timing part though, is because in certain situations, the layout positions may be calculated before the size of the nodes have stabilized. This part is WIP on my end. And I just did a quick check to see if the layout function is being invoke when hiding or unhiding and it's not so it shouldn't be blocking the render.

**Reply by MatthewAry** [reply]
Clicking the eye button and hiding/unhiding descendants is what this issue is really about.

**Reply by MatthewAry** [reply]
As for the sidenote on the provide/inject I'm doing, you're right, in most situations it would be unnecessary but I ran into some situations where I needed to do this to access the flow context in more places.

**Reply by bcakmakoglu** [reply]
>situations where I needed to do this to access the flow context in more places.

Can you describe those situations a bit more?

**Reply by bcakmakoglu** [reply]
As for the performance issue, I can notice some lag for sure although there seem to be some things that can be changed to seemingly improve this.

One issue I noticed is the passing through of props with `v-bind`.

If I write the slots out like this:
```vue
    <template #node-root="{ id, data }">
      <RootNode :id="id" v-bind="data" />
    </template>
    <template #node-silo="props">
      <BaseNode
        :can-hide-self="props.data.canHideSelf"
        :description="props.data.description"
        :label="props.data.label"
        :parent-node="props.data.parentNode"
        :can-hide-children="props.data.canHideChildren"
        :id="props.id"
        has-source
        has-target
        color="blue"
      />
    </template>

    <template #node-component="props">
      <BaseNode
        :can-hide-self="props.data.canHideSelf"
        :description="props.data.description"
        :label="props.data.label"
        :parent-node="props.data.parentNode"
        :can-hide-children="props.data.canHideChildren"
        :id="props.id"
        has-source
        has-target
        deletable
        color="red"
      />
    </template>
```

It seems to be a little bit faster 🤔

Another thing that seems to speed things up significantly is removing the vuetify tooltip buttons except for the visibility toggle.
That seems to change things *a lot* for some reason and I'm not sure if it's only related to possibly inefficient hiding of nodes.

```vue
<!-- removed all other actions -->
<VCardActions>
  <VTooltip location="bottom">
    <template #activator="{ props }">
      <VBtn v-if="children.length > 0" variant="text" icon size="small" v-bind="props" @click="toggleChildrenVisibility()">
        <VIcon :icon="childrenHidden ? mdiEye : mdiEyeOff" size="28" />
      </VBtn>
    </template>
    <span>Hide Children</span>
  </VTooltip>
</VCardActions>
```

Although I will investigate if there's some measure that I can take to get this working a bit better, maybe a CSS solution would be better but might have some drawbacks as well, we'll see ^^

**Reply by MatthewAry** [reply]
Vuetify does have some issues. They are severely understaffed.

**Reply by bcakmakoglu** [reply]
You could try using `display: none` styles yourself on nodes instead of the hidden property if that speeds things up in this case.
Although you'd have to filter the nodes during the layouting process.

**Reply by bcakmakoglu** [reply]
A small performance improvement will be released in the next minor version which should help a bit in this case :)

**Reply by MatthewAry** [reply]
> One issue I noticed is the passing through of props with v-bind.

That sounds like something upstream on the Vue level. It doesn't make sense to me why passing an object to v-bind should have any affect on perf unless there is a lot of ops needed to wire up the props in object to a component. I don't think you can use a compiler to pre-wire an object to props because you can't know ahead of time for sure what the object will contain when it's passed in at runtime. Is there any thing that documents these optimization type best practices for vue?

**Reply by MatthewAry** [reply]
> Another thing that seems to speed things up significantly is removing the vuetify tooltip buttons except for the visibility toggle.

I'll work on making a reproduction that demonstrates performance issues with VTooltip in Vuetify. If I can reproduce I will share with the Vuetify team.

**Reply by MatthewAry** [reply]
Vuetify has VTooltips prepared in an eager manner by default. This can be changed by setting `eager` to false on the component. Setting this adds a 33% perf improvement. In otherwords, `Not A Bug`

**Reply by bcakmakoglu** [reply]
Could you update to `1.35.0` and see if that helps a bit with the performance? :)
